staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	// vendor of the driver.
+	//
+	// +required
+	DriverName string `json:"driverName" protobuf:"bytes,1,name=driverName"`
+
+	// PoolName is used to identify devices. For node-local devices, this
+	// is often the node name, but this is not required.
+	//
+	// It must not be longer than 253 and must consist of one or more DNS sub-domains
+	// separated by slashes.
+	//
+	// +required
+	PoolName string `json:"poolName" protobuf:"bytes,2,name=poolName"`
+
+	// NodeName identifies the node which provides the devices.
+	// A field selector can be used to list only ResourceSlice
+	// objects belonging to a certain node.
+	//
+	// This field can be used to limit access from nodes to slices with
+	// the same node name. It also indicates to autoscalers that adding
+	// new nodes of the same type as some old node might also make new
+	// devices available.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeName *string `json:"nodeName,omitempty" protobuf:"bytes,3,opt,name=nodeName"`
+
+	// Defines which nodes have access to the devices in the pool.
+	// If the node selector is empty, all nodes have access.
+	//
+	// NodeName and NodeSelector are mutually exclusive. One of them
+	// must be set.
+	//
+	// +optional
+	NodeSelector *v1.NodeSelector `json:"nodeSelector,omitempty" protobuf:"bytes,4,opt,name=nodeSelector"`
+
+	// The generation gets bumped in all slices of a pool whenever device
+	// definitions change. A consumer must only use device definitions from slices
+	// with the highest generation number and ignore all others.
+	PoolGeneration int64 `json:"poolGeneration" protobuf:"bytes,5,name=poolGeneration"`
+
+	// The total number of slices in the pool.
+	// Consumers can use this to check whether they have
+	// seen all slices.
+	PoolSliceCount int64 `json:"poolSliceCount" protobuf:"bytes,6,name=poolSliceCount"`
+
+	// Devices lists all available devices in this pool.
+	//
+	// Must not have more than 128 entries.
+	//
+	// +required
+	// +listType=atomic
+	Devices []Device `json:"devices" protobuf:"bytes,7,name=devices"`
+
+	// FUTURE EXTENSION: some other kind of list, should we ever need it.
+	// Old clients seeing an empty Devices field can safely ignore the (to
+	// them) empty pool.
+}
+
+const ResourceSliceMaxSharedCapacity = 128
+const ResourceSliceMaxDevices = 128
+const PoolNameMaxLength = validation.DNS1123SubdomainMaxLength // Same as for a single node name.
+
+// Device represents one individual hardware instance that can be selected based
+// on its attributes.
+type Device struct {
+	// Name is unique identifier among all devices managed by
+	// the driver in the pool. It must be a DNS label.
+	//
+	// +required
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	// Attributes defines the set of attributes for this device.
+	// The name of each attribute must be unique in that set.
+	//
+	// The maximum number of attributes and capacities is 32.
```

Comment:
combined?  Or each?

Reply:
Combined. If we made it "each", our maximum size calculation is different and we would have to allow for each of them to be half of what it is now, i.e. 16 each. That's pretty small.

Side comment: the distinction between "attributes" and "capacities" still feels odd to me. As it stands now, that distinction doesn't add much value and actually removes functionality because we said that `MatchAttribute` only works for attributes, not for capacities. It's part of the plan to support partitionable devices, so if this really goes into that direction, then I'm okay with it.

/cc @johnbelamaric @klueska

Reply:
> Update the comment here, please. 

Ack.